File: bounded_readers_writers_lock.c
/* Deterministic contract lowered from the reviewed V2 domain 'bounded_readers_writers_lock'.
 * Human review of the reviewed artifact is required before trust. */

typedef struct {
    int read_count;
    _Bool writer_active;
} bounded_readers_writers_lock;

/*@
  requires \valid(counter);
  assigns counter->read_count, counter->writer_active;
  ensures counter->read_count == 0 && counter->writer_active == 0;
  ensures (0 <= counter->read_count) && (counter->read_count <= 2) && (!(counter->writer_active && (counter->read_count > 0)));
*/
void bounded_readers_writers_lock_init(bounded_readers_writers_lock *counter) {
    counter->read_count = 0;
    counter->writer_active = 0;
}

/*@
  requires \valid_read(counter);
  assigns \nothing;
  ensures \result == counter->read_count;
*/
int bounded_readers_writers_lock_get_read_count(const bounded_readers_writers_lock *counter) {
    return counter->read_count;
}

/*@
  requires \valid_read(counter);
  assigns \nothing;
  ensures \result == counter->writer_active;
*/
_Bool bounded_readers_writers_lock_get_writer_active(const bounded_readers_writers_lock *counter) {
    return counter->writer_active;
}

/*@
  requires \valid(counter);
  requires (0 <= counter->read_count) && (counter->read_count <= 2) && (!(counter->writer_active && (counter->read_count > 0)));
  requires !(counter->writer_active);
  requires counter->read_count < 2;
  assigns counter->read_count;
  ensures counter->read_count == \old(counter->read_count) + 1;
  ensures (0 <= counter->read_count) && (counter->read_count <= 2) && (!(counter->writer_active && (counter->read_count > 0)));
*/
void bounded_readers_writers_lock_reader_enter(bounded_readers_writers_lock *counter) {
    int pre_read_count = counter->read_count;
    counter->read_count = pre_read_count + 1;
}

/*@
  requires \valid(counter);
  requires (0 <= counter->read_count) && (counter->read_count <= 2) && (!(counter->writer_active && (counter->read_count > 0)));
  requires counter->read_count > 0;
  assigns counter->read_count;
  ensures counter->read_count == \old(counter->read_count) - 1;
  ensures (0 <= counter->read_count) && (counter->read_count <= 2) && (!(counter->writer_active && (counter->read_count > 0)));
*/
void bounded_readers_writers_lock_reader_exit(bounded_readers_writers_lock *counter) {
    int pre_read_count = counter->read_count;
    counter->read_count = pre_read_count - 1;
}

/*@
  requires \valid(counter);
  requires (0 <= counter->read_count) && (counter->read_count <= 2) && (!(counter->writer_active && (counter->read_count > 0)));
  requires !(counter->writer_active);
  requires counter->read_count == 0;
  assigns counter->writer_active;
  ensures counter->writer_active == 1;
  ensures (0 <= counter->read_count) && (counter->read_count <= 2) && (!(counter->writer_active && (counter->read_count > 0)));
*/
void bounded_readers_writers_lock_writer_enter(bounded_readers_writers_lock *counter) {
    counter->writer_active = 1;
}

/*@
  requires \valid(counter);
  requires (0 <= counter->read_count) && (counter->read_count <= 2) && (!(counter->writer_active && (counter->read_count > 0)));
  requires counter->writer_active;
  assigns counter->writer_active;
  ensures counter->writer_active == 0;
  ensures (0 <= counter->read_count) && (counter->read_count <= 2) && (!(counter->writer_active && (counter->read_count > 0)));
*/
void bounded_readers_writers_lock_writer_exit(bounded_readers_writers_lock *counter) {
    counter->writer_active = 0;
}

File: BoundedReadersWritersLock.rs
use prusti_contracts::*;

pub struct BoundedReadersWritersLock {
    pub read_count: i32,
    pub writer_active: bool,
}

impl BoundedReadersWritersLock {
    #[ensures(result.read_count == 0 && result.writer_active == false && (0 <= result.read_count) && (result.read_count <= 2) && (!(result.writer_active && (result.read_count > 0))))]
    pub fn new() -> Self {
        Self { read_count: 0, writer_active: false }
    }

    #[pure]
    #[ensures(result == self.read_count)]
    pub fn get_read_count(&self) -> i32 {
        self.read_count
    }

    #[pure]
    #[ensures(result == self.writer_active)]
    pub fn get_writer_active(&self) -> bool {
        self.writer_active
    }

    #[requires((0 <= self.read_count) && (self.read_count <= 2) && (!(self.writer_active && (self.read_count > 0))))]
    #[requires(!(self.writer_active))]
    #[requires(self.read_count < 2)]
    #[ensures(self.read_count == old(self.read_count) + 1)]
    #[ensures((0 <= self.read_count) && (self.read_count <= 2) && (!(self.writer_active && (self.read_count > 0))))]
    pub fn reader_enter(&mut self) {
        let pre_read_count = self.read_count;
        self.read_count = pre_read_count + 1;
    }

    #[requires((0 <= self.read_count) && (self.read_count <= 2) && (!(self.writer_active && (self.read_count > 0))))]
    #[requires(self.read_count > 0)]
    #[ensures(self.read_count == old(self.read_count) - 1)]
    #[ensures((0 <= self.read_count) && (self.read_count <= 2) && (!(self.writer_active && (self.read_count > 0))))]
    pub fn reader_exit(&mut self) {
        let pre_read_count = self.read_count;
        self.read_count = pre_read_count - 1;
    }

    #[requires((0 <= self.read_count) && (self.read_count <= 2) && (!(self.writer_active && (self.read_count > 0))))]
    #[requires(!(self.writer_active))]
    #[requires(self.read_count == 0)]
    #[ensures(self.writer_active == true)]
    #[ensures((0 <= self.read_count) && (self.read_count <= 2) && (!(self.writer_active && (self.read_count > 0))))]
    pub fn writer_enter(&mut self) {
        self.writer_active = true;
    }

    #[requires((0 <= self.read_count) && (self.read_count <= 2) && (!(self.writer_active && (self.read_count > 0))))]
    #[requires(self.writer_active)]
    #[ensures(self.writer_active == false)]
    #[ensures((0 <= self.read_count) && (self.read_count <= 2) && (!(self.writer_active && (self.read_count > 0))))]
    pub fn writer_exit(&mut self) {
        self.writer_active = false;
    }
}
